### PR TITLE
Progress bar

### PR DESCRIPTION
**Pull request by shaybeau731**
There's still some iffy behaviour happening, but I thought it'd be good for people to look this over and give any feedback they may have.

How it works so far: 
- a progress bar gets added to all code cells when running
- a progress bar has class "progress-bar-warning" and text "Loading Spark..." when it is initialized
- cell_jobs keeps track of code cells currently running, as well as their corresponding Spark jobs
- as the cache gets automatically updated (currently once per second), a trigger event goes off to update progress bars for each cell in cell_jobs
- code cells not running Spark have progress bars that remain in class "progress-bar-warning"
- when any cell is changed (eg. write something in input; something loads for output), all progress bars with class "progress-bar-warning" or "progress-bar-success" get removed

What is iffy:
- the first Spark cell that is executed can take a while to load, as the SparkContext gets up and running; since no jobs are being run at this point, I'm not sure how to monitor its progress (or if it needs monitoring at all)
- I upped the UPDATE_FREQUENCY of the cache so that progress bars can update more regularly... but this makes things very inefficient :/ 
- non-Spark cells ideally shouldn't have a progress bar... (idea just now - initially hide progress bar?)
- each SparkContext runs on a different port, so I'm not sure if you can have multiple Spark notebooks open at the same time (as we are only monitoring the 1st port)

### CODE REVIEW

**Comment by mreid-moz** [comment]
I've tried this out - it's really cool.

> The first Spark cell that is executed can take a while to load, as the SparkContext gets up and running; since no jobs are being run at this point, I'm not sure how to monitor its progress (or if it needs monitoring at all)

We could keep a variable that gets initialized the first time we successfully hit the Spark API.

> I upped the UPDATE_FREQUENCY of the cache so that progress bars can update more regularly... but this makes things very inefficient :/

I think we can leave the update freq at 5 or 10 seconds if we add code to also update the cache whenever a cell starts executing - that way we'd get the latest data right when we need it, but wouldn't need to hammer the API all the time (although once per second may not be too bad anyways).

>  Non-Spark cells ideally shouldn't have a progress bar... (idea just now - initially hide progress bar?)

+1 for initially hiding the progress bar, then show & update if and only if we detect a running Spark Job

> Each SparkContext runs on a different port, so I'm not sure if you can have multiple Spark notebooks open at the same time (as we are only monitoring the 1st port)

I'm not too worried about this (at least not yet).

A couple of other things. I noticed that the progress bar doesn't always reflect an active job. I think it's just grabbing the first job from the cache, which is often the previous completed job. Forcing a cache refresh when a cell starts executing as described above may fix this.

If you queue up a few cells to run, the progress bar sometimes shows up in the wrong place. I think we need to track which cells are _pending_ vs. _actively executing_ and maybe have another state for the progress bar for "waiting to run" rather than "loading spark".  Any currently running spark jobs should affect the currently executing code cell's progress bar, and there should only be one executing cell at a time (I think that's how it works, at least).  

A single cell ca  run multiple jobs too (think a for loop or just multiple operations in a row), and it seems like it would be overly complicated to try to add them all up and maintain "overall" progress across an arbitrary number jobs. I think the progress bar should reflect the currently-running job (and just reset when a new job starts). If we could include a counter of "completed jobs for this cell", that would be a bonus.

**Comment by shaybeau731** [comment]
I made a few of the changes you suggested, and now things are running much smoother!

- cells to be executed are now stored in a queue
- only the currently executing cell will have a progress bar (and hence, only one progress bar is updated at a time)
- a progress bar is hidden until a new job is detected
- a progress bar reflects the status of the most recent job 
- a counter above the progress bar indicates how many jobs have been executed in the current cell
- when a cell is finished executing, its progress bar is removed, and a progress bar will be added to the next cell in the queue (if it isn't empty)
- the cache updates more frequently when a cell is executing

Please let me know what you think!

**Comment by frol** [comment]
I have also tried out your patchset. Works fine enough for development, but some glitches occur here and there:

* If I run several cells at a time, a progressbar may appear on any of the running cells.
* Another glitch when executing several cells at once is that the progressbar end up stuck in "processing" state even after all cells are already done.
* Once the first glitch occurred, progressbar will appear in random cells even on one-cell executions.

Here is a video-recording of the described glitches: https://youtu.be/H6-Ekrl66wM

Thank you!

**Comment by mreid-moz** [comment]
@musicaljelly @yeah568 @agjwong Please also have a look.  Thanks!

**Comment by mreid-moz** [comment]
@frol thanks for the feedback!

**Comment by musicaljelly** [comment]
Runs fine on my machine, and the code looks good to me!

**Comment by musicaljelly** [comment]
Should we bump the version number?

File: jupyter-spark/extensions/spark.js [hunk 1]
```diff
@@ -1,5 +1,8 @@
+
 var API = "/spark/api/v1";
 var UPDATE_FREQUENCY = 10000; // ms
+var UPDATE_FREQUENCY_ACTIVE = 1000;
+var PROGRESS_COUNT_TEXT = "Completing Spark job ";
```
Comment (mreid-moz) [inline jupyter-spark/extensions/spark.js]: I think "Running" would be a better word here than "Completing".

File: jupyter-spark/extensions/spark.js [hunk 2]
```diff
@@ -24,17 +35,26 @@ var update_cache = function(callbacks) {
...
-            $.getJSON(API + '/applications/' + application.id + '/jobs').done(function (jobs) {
-                cache[i] = application;
-                cache[i].jobs = jobs;
-
-                num_completed++;
-                if (num_completed === num_applications && cbs) {
-                    cbs.fire(cache);
-                }
+        // Check if Spark is running before processing applications
+        if(!applications.hasOwnProperty('error')){
```
Comment (mreid-moz) [inline jupyter-spark/extensions/spark.js]: should set `spark_is_running = false` if it *does* have the error property (ie. if Spark crashes or is shut down)

**Comment by mreid-moz** [comment]
I'm still seeing the progress bar attaching to unexpected cells - I'll dig into it more soon.

**Comment by yeah568** [comment]
Apart from the issues already mentioned, seems to be working fine here.

**Comment by shaybeau731** [comment]
Thanks for the feedback! And thank you @frol for taking the time to make a video, it was a big help!
I made some changes in how progress bars get removed, and that bug should be fixed now!

**Comment by mreid-moz** [comment]
I just tried it again, and it looks like things work great! Thanks!

File: jupyter-spark/extensions/spark.js [hunk 3]
```diff
@@ -1,5 +1,8 @@
+
```
Comment (mreid-moz) [inline jupyter-spark/extensions/spark.js]: nit: remove blank line

File: jupyter-spark/extensions/spark.js [hunk 4]
```diff
@@ -8,8 +11,16 @@ application.jobs is the result of the /applications/applicationId/jobs
...
+var current_update_frequency;
+
+var spark_is_running = false;
+var cell_queue = [];
+var current_cell;
+var cell_jobs_counter = 0;
+var jobs_in_cache = 0;
 
 var update = function() {
+    //console.log("updating cache");
```
Comment (mreid-moz) [inline jupyter-spark/extensions/spark.js]: remove commented-out code

File: jupyter-spark/extensions/spark.js [hunk 5]
```diff
@@ -24,20 +35,32 @@ var update_cache = function(callbacks) {
...
+                    if (jobs.length > jobs_in_cache && cell_queue.length > 0 ) {
+                        $(document).trigger('update.progress.bar');
+                    };
+                });
             });
-        });
+        } else {
+            spark_is_running = false;
+        }
+
```
Comment (mreid-moz) [inline jupyter-spark/extensions/spark.js]: nit: remove empty line

File: jupyter-spark/extensions/spark.js [hunk 6]
```diff
@@ -24,20 +35,32 @@ var update_cache = function(callbacks) {
...
     });
 };
 
+
```
Comment (mreid-moz) [inline jupyter-spark/extensions/spark.js]: nit: remove 2nd empty line

File: jupyter-spark/extensions/spark.js [hunk 7]
```diff
@@ -99,15 +134,20 @@ var create_table_row = function(e) {
...
     progress_bar_div.append(progress_bar);
-    row.append($('<td/>').append(progress_bar_div));
-    return row;
+    return progress_bar_div;
 };
 
 
-define(['jquery', 'base/js/dialog'], function ($, dialog) {
+define(['jquery', 'base/js/dialog', 'base/js/events', 'notebook/js/codecell'], function ($, dialog, events, codecell) {
+
```
Comment (mreid-moz) [inline jupyter-spark/extensions/spark.js]: nit: remove blank line

File: jupyter-spark/extensions/spark.js [hunk 8]
```diff
@@ -121,7 +161,116 @@ define(['jquery', 'base/js/dialog'], function ($, dialog) {
...
+        if (is_spark_cell(cell)) {
+            window.clearInterval(current_update_frequency);
+            current_update_frequency = window.setInterval(update, UPDATE_FREQUENCY_ACTIVE);
+            cell_queue.push(cell);
+            current_cell = cell_queue[0];
+            add_progress_bar(current_cell);
+        };
+    };
+
+
```
Comment (mreid-moz) [inline jupyter-spark/extensions/spark.js]: nit: remove one of these blank lines

File: jupyter-spark/extensions/spark.js [hunk 9]
```diff
@@ -121,7 +161,116 @@ define(['jquery', 'base/js/dialog'], function ($, dialog) {
...
+            var progress_bar_container = $('<div/>')
+                .addClass('progress-container')
+                .css({'border': 'none', 'border-top': '1px solid #CFCFCF'})
+
+            progress_bar = create_progress_bar('progress-bar-warning', 1, 5);
+            progress_bar.hide();
+            progress_bar.appendTo(progress_bar_container);
+            jobs_completed_container.appendTo(input_area);
+            progress_bar_container.appendTo(input_area);
+
```
Comment (mreid-moz) [inline jupyter-spark/extensions/spark.js]: nit: remove one of these blank lines

File: jupyter-spark/extensions/spark.js [hunk 10]
```diff
@@ -121,7 +161,116 @@ define(['jquery', 'base/js/dialog'], function ($, dialog) {
...
+        };
+    };
+
+    var is_spark_cell = function(cell) {
+        // TODO: Find a way to detect if cell is actually running Spark
+        return (cell instanceof CodeCell)
+    };
+
     var load_ipython_extension = function () {
+
```
Comment (mreid-moz) [inline jupyter-spark/extensions/spark.js]: nit: remove blank line

**Comment by mreid-moz** [comment]
Ok, I added a few style / whitespace nitpicks, but otherwise this looks great!

**Comment by mreid-moz** [comment]
Resolves #3

**Comment by frol** [comment]
Thank you guys for your work! I was dreaming of this integration since I run my first code on Spark...

**Pull request merged**